addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HAConfigTypeAdapterFactory.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.mqtt.generic.internal.convention.homeassistant;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+import com.google.gson.Gson;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * This a Gson type adapter factory.
+ *
+ * It will create a type adapter for every class derived from {@link HAConfiguration} and ensures,
+ * that abbreviated names are replaces with their long versions during the read.
+ *
+ * In elements, whose name end in'_topic' '~' replacement is performed.
+ *
+ * The adapters also handle {@link HAConfiguration.Device}
+ *
+ * @author Jochen Klein- Initial contribution
+ */
+
+@NonNullByDefault
+public class HAConfigTypeAdapterFactory implements TypeAdapterFactory {
+
+    public final static HAConfigTypeAdapterFactory INSTANCE = new HAConfigTypeAdapterFactory();
```

Comment:
Singletons are bad design in general, you know that probably (bad for testing!).
But specifically in Java: They are not garbage collected until the classloader get's unloaded. Not so a problem in a regular application, as that is the requested behaviour. But we are in an OSGi environment and bundles keep coming and going over time and the classloader gets unloaded only when the OSGi container shuts down afaik.

Can't we use static methods instead? The singleton instance doesn't seem to have a state anyway.

Reply:
Yes, this evolving code. The Type adaoter is (now) only used in one place, so it can be constructed there.